nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java

```diff
@@ -0,0 +1,376 @@
+    }
+
+    @Override
+    public void start() {
+        final RetryPolicy retryPolicy = new RetryForever(5000);
+        curatorClient = CuratorFrameworkFactory.newClient(zkClientConfig.getConnectString(),
+            zkClientConfig.getSessionTimeoutMillis(), zkClientConfig.getConnectionTimeoutMillis(), retryPolicy);
+        curatorClient.start();
+
+        this.future = flowEngine.scheduleWithFixedDelay(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    monitorHeartbeats();
+                } catch (final Exception e) {
+                    clusterCoordinator.reportEvent(null, Severity.ERROR, "Failed to process heartbeats from nodes due to " + e.toString());
+                    logger.error("Failed to process heartbeats", e);
+                }
+            }
+        }, heartbeatIntervalMillis, heartbeatIntervalMillis, TimeUnit.MILLISECONDS);
+    }
+
+    private CuratorFramework getClient() {
+        return curatorClient;
+    }
+
+    @Override
+    public void stop() {
+        final CuratorFramework client = getClient();
+        if (client != null) {
+            client.close();
+        }
+
+        if (future != null) {
+            future.cancel(true);
+        }
+    }
+
+    @Override
+    public NodeHeartbeat getLatestHeartbeat(final NodeIdentifier nodeId) {
+        return latestHeartbeatMessages.get(nodeId);
+    }
+
+
+    /**
+     * Fetches all of the latest heartbeats from ZooKeeper
+     * and updates the Cluster Coordinator as appropriate,
+     * based on the heartbeats received.
+     *
+     * Visible for testing.
+     */
+    synchronized void monitorHeartbeats() {
+        final StopWatch fetchStopWatch = new StopWatch(true);
+        final Map<NodeIdentifier, NodeHeartbeat> latestHeartbeats = fetchHeartbeats();
+        if (latestHeartbeats == null || latestHeartbeats.isEmpty()) {
+            // failed to fetch heartbeats from ZooKeeper; don't change anything.
+            clusterCoordinator.reportEvent(null, Severity.WARNING, "Failed to retrieve any new heartbeat information for nodes from ZooKeeper. "
+                + "Will not make any decisions based on heartbeats.");
+            return;
+        }
+
+        this.latestHeartbeatMessages = new HashMap<>(latestHeartbeats);
+        fetchStopWatch.stop();
+
+        final StopWatch procStopWatch = new StopWatch(true);
+        for (final NodeHeartbeat heartbeat : latestHeartbeats.values()) {
+            try {
+                processHeartbeat(heartbeat);
+            } catch (final Exception e) {
+                clusterCoordinator.reportEvent(null, Severity.ERROR,
+                    "Received heartbeat from " + heartbeat.getNodeIdentifier() + " but failed to process heartbeat due to " + e);
+                logger.error("Failed to process heartbeat from {} due to {}", heartbeat.getNodeIdentifier(), e.toString());
+                logger.error("", e);
```

Comment:
Unfortunately, the SLF4J api doesn't provide an 'error(Throwable)' method - you must provide a message.